builtin/blame.c

```diff
@@ -902,6 +902,8 @@ int cmd_blame(int argc, const char **argv, const char *prefix)
 	long anchor;
 	const int hexsz = the_hash_algo->hexsz;
 
+	prepare_repo_settings(the_repository);
```

Comment:
Comment on commit message:

> blame: enable and test the sparse index
> 
> Enable the sparse index for the 'git blame' command. The index was already
> not expanded with this command, so the most interesting thing to do is
> to add tests that verify that 'git blame' behaves
> correctly when the sparse index is enabled and that its performance
> improves. More specifically, these cases are...

I re-wrapped this paragraph (in `vim`, use `SHIFT+V` to go to visual mode, select the lines you care about, then type `gq` to wrap the lines). I also terminated the paragraph in a colon.

```
Enable the sparse index for the 'git blame' command. The index was
already not expanded with this command, so the most interesting thing to
do is to add tests that verify that 'git blame' behaves correctly when
the sparse index is enabled and that its performance improves. More
specifically, these cases are:
```

> 1. The index is not expanded for `blame` when given paths in the sparse
> checkout cone at multiple levels.
> 2. Performance measurably improves for `blame` with sparse index when
> given paths in the sparse checkout cone at multiple levels.

Style nit: I prefer a full line between bullets and left-alignment to match the first character after the bullet:

```
1. The index is not expanded for `blame` when given paths in the sparse
   checkout cone at multiple levels.

2. Performance measurably improves for `blame` with sparse index when
   given paths in the sparse checkout cone at multiple levels.
```
 
> The following are the performance results comparing the branch with
> sparse index enabled for `blame` with the vfs-2.33.0 branch in the
> microsoft/git repository:
> 
> ```
> Test                                       msft/vfs-2.33.0   sparse-index-blame
> -----------------------------------------------------------------------------------
> 2000.2: git blame f2/f4/a (full-v3)        0.31(0.21+0.09)   0.32(0.21+0.10) +3.2%
> 2000.3: git blame f2/f4/a (full-v4)        0.29(0.19+0.08)   0.31(0.21+0.08) +6.9%
> 2000.4: git blame f2/f4/a (sparse-v3)      0.55(0.38+0.14)   0.23(0.14+0.07) -58.2%
> 2000.5: git blame f2/f4/a (sparse-v4)      0.57(0.40+0.16)   0.23(0.15+0.07) -59.6%
> 2000.6: git blame f2/f4/f3/a (full-v3)     0.77(0.56+0.20)   0.85(0.60+0.22) +10.4%
> 2000.7: git blame f2/f4/f3/a (full-v4)     0.78(0.56+0.20)   0.81(0.59+0.21) +3.8%
> 2000.8: git blame f2/f4/f3/a (sparse-v3)   1.07(0.78+0.28)   0.72(0.52+0.19) -32.7%
> 2000.9: git blame f2/f4/f3/a (sparse-v4)   1.05(0.78+0.26)   0.73(0.51+0.19) -30.5%
> ```

There are a couple things that will look strange to a casual reader here:

1. The rows are numbered, and those numbers won't match the expectations of someone running the full script. Could you replace the numbers with the ones they would actually have if you ran the full script (without commenting out other tests)?

2. These second set of numbers made me think you ran the test with an increased depth compared to the others (the `test_seq 1 4` trick). Is that the case? Or, do we really have such a drastic performance change when asking for a slightly deeper path?

> We do not include paths outside the sparse checkout cone because blame
> currently does not support blaming files outside of the sparse
> definition.

Oh? Excellent. Does it just complain that the file doesn't exist, or specifically that it is marked with the `SKIP_WORKTREE` bit? It might be helpful to add details for this exact warning/error, for context.

Reply:
I think I have addressed everything here except the second question re: things that will look strange to a casual reader. I did not do the `test_seq 1_4` trick, so if that is something that warrants further investigation, I can definitely dig deeper on Monday.

Reply:
I did a performance trace and the deeper path _is_ spending more time in the actual history walk, primarily related to rename detection. This isn't anything to worry about. Thanks.